src/Translation_unit.mli

```diff
@@ -33,34 +32,34 @@ type 'a t =
   ; normalize: Conf.t -> 'a with_comments -> 'a
   ; printast: Caml.Format.formatter -> 'a -> unit }
 
-(** Existential package of a type of translation unit and its operations. *)
-type x = XUnit : 'a t -> x
-
-exception Warning50 of (Location.t * Warnings.t) list
-
-type result =
-  | Ok
-  | Invalid_source of exn
-  | Unstable of int
-  | Ocamlformat_bug of exn
+type error =
+  | Invalid_source of {input_name: string; exn: exn}
+  | Unstable of
+      { input_name: string
+      ; iteration: int
+      ; prev: string
+      ; next: string }
+  | Ocamlformat_bug of {input_name: string; exn: exn}
   | User_error of string
 
-val parse :
-     (Lexing.lexbuf -> 'a)
-  -> Conf.t
-  -> input_file:string
-  -> In_channel.t
-  -> 'a with_comments
+val parse : (Lexing.lexbuf -> 'a) -> Conf.t -> string -> 'a with_comments
 
-val parse_print :
-     x
+val format :
+     'a t
   -> Conf.t
+  -> ?dump_ast:(suffix:string -> 'a -> unit)
+  -> ?dump_formatted:(suffix:string -> string -> string option)
   -> input_name:string
-  -> input_file:string
-  -> In_channel.t
-  -> string option
-  -> result
-(** [parse_print xunit conf input_name input_file input_channel output_file]
-    parses the contents of [input_channel], using [input_name] for error
-    messages, and referring to the contents of [input_file] to improve
-    comment placement, and writes formatted code into [output_file]. *)
+  -> source:string
+  -> parsed:('a with_comments, exn) Result.t
+  -> unit
+  -> (string, error) Result.t
```

Comment:
The only thing a client can do with the returned error is pass it to `print_error`, right? So what about partially applying `print_error` to the error and returning the `print_error` function instead?

Reply:
A client could deconstruct the error manually.

Reply:
Yes, true. I'm just wondering whether that is more functionality than a client is going to want, and hence maybe the narrower interface would be preferable. I have no strong view either way, just mentioning it.